core/src/main/java/hudson/model/Run.java

```diff
@@ -551,8 +553,18 @@ public boolean isLogUpdated() {
      * @since 1.257
      */   
     public final @Nonnull Charset getCharset() {
-        if(charset==null)   return Charset.defaultCharset();
-        return Charset.forName(charset);
+        if (charsetInstance==null)
+            return StandardCharsets.UTF_8;
+        return charsetInstance;
+    }
+
+    /**
+     * Sets the charset in which the log file is written.
+     * @since TODO
+     */
+    public final void setCharset(Charset charset) {
+        this.charsetInstance = charset;
```

Comment:
Not enough. `onLoad` must also restore `charsetInstance` based on the persisted `charset`, or `getCharset` will wind up always returning UTF-8 after a restart.